### PR TITLE
Add temporal navigation step for "source timestamps"

### PR DESCRIPTION
**Pull request by nyalldawson**
(Includes https://github.com/qgis/QGIS/pull/42399, and by extension includes #42372)

When selected, this causes the temporal navigation to step between all available time ranges from layers in the project.

It's useful when a project contains layers with non-contiguous available times, e.g. from a WMS-T which images available at
irregular dates, and you want to only step between time ranges where the next available image is shown.

Refs Natural resources Canada Contract: 3000720707

The screencast below shows this in action -- note how after switching to the "source timestamps" mode the time jumps become irregular steps, corresponding to each available image time from the loaded WMST services

Peek 2021-03-24 15-15

### CODE REVIEW

**Comment by elpaso** [comment]
@nyalldawson 

```patch
diff --git a/src/providers/postgres/raster/qgspostgresrasterprovider.cpp b/src/providers/postgres/raster/qgspostgresrasterprovider.cpp
index 89d4044af3..17614554ec 100644
--- a/src/providers/postgres/raster/qgspostgresrasterprovider.cpp
+++ b/src/providers/postgres/raster/qgspostgresrasterprovider.cpp
@@ -1432,6 +1432,27 @@ bool QgsPostgresRasterProvider::initFieldsAndTemporal( )
                                          QStringLiteral( "PostGIS" ), Qgis::Warning );
             }
           }
+          // Set temporal ranges
+          QList< QgsDateTimeRange > allRanges;
+          const QString sql =  QStringLiteral( "SELECT DISTINCT %1::timestamp "
+                                               "FROM %2 %3 ORDER BY %1::timestamp" ).arg( quotedIdentifier( temporalFieldName ),
+                                                   mQuery,
+                                                   where );
+
+          QgsPostgresResult result( connectionRO()->PQexec( sql ) );
+          if ( PGRES_TUPLES_OK == result.PQresultStatus() && result.PQntuples() > 0 )
+          {
+            for( qlonglong row = 0; row < result.PQntuples(); ++row )
+            {
+              allRanges.push_back( QgsDateTimeRange { QDateTime::fromString( result.PQgetvalue( row, 0 ), Qt::DateFormat::ISODate ) } );
+            }
+            temporalCapabilities()->setAllAvailableTemporalRanges( allRanges );
+          }
+          else
+          {
+            QgsMessageLog::logMessage( tr( "No temporal ranges detected in raster temporal capabilities for field %1: %2" ).arg( temporalFieldName, mUri.param( QStringLiteral( "temporalDefaultTime" ) ) ),
+                                       QStringLiteral( "PostGIS" ), Qgis::Info);
+          }
         }
         else
         {
```

**Comment by elpaso** [comment]
@nyalldawson there is still something that doesn't work for me: when setting the "Default time" in temporal capabilities widget I would expect to see a list of ranges instead of the calendar widget (but I may be missing something).

Without knowing the ranges a that stage it is almost impossible to get them right.

Nice job btw: it is working good!

**Comment by nyalldawson** [comment]
@elpaso is that with your patch?

**Comment by elpaso** [comment]
> @elpaso is that with your patch?

No, the patch is only about collecting the ranges for PG rasters and make this PR (the irregular step) actually work with PG raster.

**Comment by nyalldawson** [comment]
@elpaso can you share the layer source and a screencast?

**Comment by elpaso** [comment]
> @elpaso can you share the layer source and a screencast?

done in PM

**Comment by nyalldawson** [comment]
@elpaso your issue is addressed in https://github.com/qgis/QGIS/pull/42455

File: src/core/qgsunittypes.cpp [hunk 1]
```diff
@@ -1275,6 +1275,8 @@ QString QgsUnitTypes::encodeUnit( QgsUnitTypes::TemporalUnit unit )
       return QStringLiteral( "dec" );
     case TemporalCenturies:
       return QStringLiteral( "c" );
+    case TemporalIrregularStep:
+      return QStringLiteral( "xxx" );
```
Comment (nirvn) [inline src/core/qgsunittypes.cpp:1279]: xxx, was that meant to be a temporary string prior to it being replaced by something better?

**Review by nirvn — APPROVED** [review]
Just one question.

File: src/core/qgsunittypes.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsunittypes.cpp:1279]: No, it's deliberate. It's used when writing /reading the unit to projects and I wanted something which would never clash with any other units we may add in future :)

**Pull request merged**